File: problems/LeetCode/double_pointers/threeSum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []

        for i in range(n - 2):
            x = nums[i]

            # skip duplicates 
            if i > 0 and x == nums[i-1]:
                continue
            
            # optimization
            if x + nums[i+1] + nums[i+2] > 0:
                break
            
            if x + nums[-2] + nums[-1] < 0:
                continue
            
            j = i + 1
            k = n - 1

            while j < k:
                sum = x + nums[j] + nums[k]
                if sum > 0:
                    k -= 1
                elif sum < 0:
                    j += 1
                else:
                    ans.append([x, nums[j], nums[k]])
                    j += 1
                    # skip duplicates
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                    k -= 1
                    while k > j and nums[k] == nums[k+1]:
                        k -= 1
        return ans
                   

File: problems/LeetCode/TopInterview150/threeSum.ts
function threeSum(nums: number[]): number[][] {
  let sorted = nums.sort((a, b) => a - b);
  let ans = [] as number[][];

  for (let i = 0; i < sorted.length; i++) {
    // Remove duplicates
    if (i !== 0 && sorted[i] === sorted[i - 1]) {
      continue;
    }
    twoSum(sorted, i + 1, -sorted[i], ans);
  }

  return ans;
}

function twoSum(
  sorted: number[],
  left: number,
  target: number,
  ans: number[][]
) {
  let right = sorted.length - 1;

  while (left < right) {
    if (sorted[left] + sorted[right] === target) {
      ans.push([-target, sorted[left], sorted[right]]);
      left++;
      right--;

      // Remove duplicates
      while (left < right && sorted[left] === sorted[left - 1]) {
        left++;
      }

      while (left < right && sorted[right] === sorted[right + 1]) {
        right--;
      }
    } else if (sorted[left] + sorted[right] < target) {
      left++;
    } else {
      right--;
    }
  }
}

// Time complexity: O(n^2) +O(nlog(n)) = O(n^2)
// Space complexity: O(n)
